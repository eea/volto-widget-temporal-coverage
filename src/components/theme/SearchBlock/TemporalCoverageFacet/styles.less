.temporal-facet {
  .ui.header {
    padding-bottom: 0;
    margin-bottom: 0;
  }
  .years-input {
    display: flex;
    width: 100%;
    flex-direction: row;
    justify-content: flex-start;
    padding: 0;
    margin: 0;
    .right {
      display: flex;
      margin-left: auto;
    }
  }
}
