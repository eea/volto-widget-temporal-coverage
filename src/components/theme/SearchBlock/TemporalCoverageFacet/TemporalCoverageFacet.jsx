import React from 'react';
import { Header, Input } from 'semantic-ui-react';

import './styles.less';

import { Slider, Rail, Handles, Tracks, Ticks } from 'react-compound-slider';
import { SliderRail, Handle, Track, Tick } from './SliderComponents';
import debounce from 'lodash.debounce';

const getRangeStartEnd = (range) => {
  return {
    min: Math.min.apply(
      Math,
      range
        ?.filter((year) => parseInt(year.value) > 0)
        ?.map((year) => parseInt(year.value)),
    ),
    max: Math.max.apply(
      Math,
      range
        ?.filter((year) => parseInt(year.value) > 0)
        ?.map((year) => parseInt(year.value)),
    ),
  };
};

const sliderStyle = {
  position: 'relative',
  width: '100%',
};

const useDebouncedInput = (initialValue) => {
  const [storedValue, setValue] = React.useState(initialValue);
  React.useEffect(() => {
    if (initialValue !== storedValue && storedValue === undefined) {
      // console.log('should update internal state', {
      //   initialValue,
      //   storedValue,
      // });
      setValue(initialValue);
    }
  }, [storedValue, initialValue]);

  const timerRef = React.useRef();

  const changeHandler = React.useCallback((newValue, callback) => {
    setValue(newValue);
    timerRef.current && clearTimeout(timerRef.current);
    timerRef.current = setTimeout(() => {
      // console.log('calling callback');
      callback();
    }, 1000);
  }, []);

  return [storedValue, changeHandler];
};

const TemporalCoverageFacet = (props) => {
  const { facet, onChange, choices, value } = props;
  const { min, max } = getRangeStartEnd(choices);

  const [start, end] =
    value?.length > 0
      ? [
          parseInt(value[0] ? value[0] : min),
          parseInt(value[value?.length - 1] ? value[value?.length - 1] : max),
        ]
      : [min, max];

  const [startValue, handleChangeStartValue] = useDebouncedInput(start);

  const [endValue, handleChangeEndValue] = useDebouncedInput(end);

  const debouncedSliderChangeHandler = debounce((sliderRange) => {
    if (
      !(
        sliderRange?.[0] >= min &&
        sliderRange?.[1] <= max &&
        sliderRange?.[0] <= max &&
        sliderRange?.[1] >= min
      )
    ) {
      return;
    }
    onChange(facet?.field?.value, [sliderRange[0], sliderRange[1]]);
    if (sliderRange[0] !== startValue)
      handleChangeStartValue(parseInt(sliderRange[0]), () => {});
    if (sliderRange[0] !== endValue)
      handleChangeEndValue(parseInt(sliderRange[1]), () => {});
  }, 300);

  return (
    <div className="temporal-facet">
      <Header as="h4">{facet?.title ?? facet?.field?.label}</Header>
      <div className="ui form years-input">
        <Input
          type="number"
          value={startValue}
          onChange={(e, { value }) => {
            handleChangeStartValue(parseInt(value), () => {
              if (parseInt(value) >= min && parseInt(value) <= max) {
                onChange(facet?.field?.value, [parseInt(value), endValue]);
              }
            });
          }}
          min={min}
          max={max}
        />
        <Input
          type="number"
          className="right"
          value={endValue}
          onChange={(e, { value }) => {
            handleChangeEndValue(parseInt(value), () => {
              if (parseInt(value) >= min && parseInt(value) <= max) {
                onChange(facet?.field?.value, [startValue, parseInt(value)]);
              }
            });
          }}
          min={min}
          max={max}
        />
      </div>
      <Slider
        step={1}
        domain={[min, max]}
        rootStyle={sliderStyle}
        onChange={(e) => {
          debouncedSliderChangeHandler(e);
        }}
        values={[start, end]}
      >
        <Rail>
          {({ getRailProps }) => <SliderRail getRailProps={getRailProps} />}
        </Rail>
        <Handles>
          {({ handles, activeHandleID, getHandleProps }) => (
            <div className="slider-handles">
              {handles.map((handle) => (
                <Handle
                  key={handle.id}
                  handle={handle}
                  domain={[min, max]}
                  isActive={handle.id === activeHandleID}
                  getHandleProps={getHandleProps}
                />
              ))}
            </div>
          )}
        </Handles>
        <Tracks left={false} right={false}>
          {({ tracks, getTrackProps }) => (
            <div className="slider-tracks">
              {tracks.map(({ id, source, target }) => (
                <Track
                  trackColor={'#00548a'}
                  key={id}
                  source={source}
                  target={target}
                  getTrackProps={getTrackProps}
                />
              ))}
            </div>
          )}
        </Tracks>
        <Ticks count={5}>
          {({ ticks }) => (
            <div className="slider-ticks">
              {ticks.map((tick) => (
                <Tick key={tick.id} tick={tick} count={ticks.length} />
              ))}
            </div>
          )}
        </Ticks>
      </Slider>
    </div>
  );
};

TemporalCoverageFacet.stateToValue = ({
  facetSettings,
  index,
  selectedValue,
}) => {
  return selectedValue || [null, null];
};

TemporalCoverageFacet.valueToQuery = ({ value, facet }) => {
  const years = [parseInt(value?.[0]), parseInt(value?.[value.length - 1])];
  let allYears = Array.from(
    { length: years?.[1] - years?.[0] + 1 },
    (_, index) => (years?.[0] + index).toString(),
  );
  return value
    ? {
        i: facet?.field?.value,
        o: 'plone.app.querystring.operation.list.contains',
        v: allYears,
      }
    : null;
};

export default TemporalCoverageFacet;
