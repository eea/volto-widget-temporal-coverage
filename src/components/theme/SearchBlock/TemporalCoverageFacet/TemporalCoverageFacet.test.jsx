import React from 'react';
import { render, screen } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom/extend-expect';

import TemporalCoverageFacetFilterListEntry from './TemporalCoverageFacetFilterListEntry';
import { Tick } from './SliderComponents';
import { Ticks } from 'react-compound-slider';
import TemporalCoverageFacet from './TemporalCoverageFacet';

const mockStore = configureStore([thunk]);

it('renders TemporalCoverageFacet', () => {
  const store = mockStore({
    intl: {
      locale: 'en',
      messages: {},
    },
    temporal: [
      { value: '1890', label: '1890' },
      { value: '1900-1905', label: '1900-1905' },
      { value: '1915', label: '1915' },
    ],
  });

  const { container } = render(
    <Provider store={store}>
      <TemporalCoverageFacet
        choices={[
          { value: '1890', label: '1890' },
          { value: '1900-1905', label: '1900-1905' },
          { value: '1915', label: '1915' },
        ]}
      />
    </Provider>,
  );
  expect(
    container.querySelector('.temporal-facet .ui.form.years-input .ui.input'),
  ).toBeInTheDocument();
  expect(
    container.querySelector(
      '.temporal-facet .ui.form.years-input .ui.input.right',
    ),
  ).toBeInTheDocument();

  expect(
    container.querySelector('input[max="1915"][min="1890"]'),
  ).toBeInTheDocument();
  expect(
    container.querySelector('.ui.input input[max="1915"][min="1890"]'),
  ).toHaveValue(1890);
  expect(
    container.querySelector('.ui.input.right input[max="1915"][min="1890"]'),
  ).toHaveValue(1915);

  expect(container.querySelector('.slider-handles')).toBeInTheDocument();
  expect(
    container.querySelector(
      '.slider-handles div[aria-valuemax="1915"][aria-valuemin="1890"][aria-valuenow="1890"][role="slider"]',
    ),
  ).toBeInTheDocument();
  expect(
    container.querySelector(
      '.slider-handles div[aria-valuemax="1915"][aria-valuemin="1890"][aria-valuenow="1915"][role="slider"]',
    ),
  ).toBeInTheDocument();

  expect(container.querySelector('.slider-ticks')).toBeInTheDocument();
  expect(screen.getByText(1890)).toHaveStyle('left: 0%');
  expect(screen.getByText(1895)).toHaveStyle('left: 20%');
  expect(screen.getByText(1900)).toHaveStyle('left: 40%');
  expect(screen.getByText(1905)).toHaveStyle('left: 60%');
  expect(screen.getByText(1910)).toHaveStyle('left: 80%');
  expect(screen.getByText(1915)).toHaveStyle('left: 100%');
});

it('renders TemporalCoverageFacetFilterListEntry', () => {
  const store = mockStore({
    intl: {
      locale: 'en',
      messages: {},
    },
    temporal: [
      { value: '1890', label: '1890' },
      { value: '1900-1905', label: '1900-1905' },
      { value: '1915', label: '1915' },
    ],
  });

  const { container } = render(
    <Provider store={store}>
      <TemporalCoverageFacetFilterListEntry
        facets={{ temporal_coverage: [1890, 1915] }}
        facet={'temporal_coverage'}
        choices={[
          { value: '1890', label: '1890' },
          { value: '1900-1905', label: '1900-1905' },
          { value: '1915', label: '1915' },
        ]}
      />
    </Provider>,
  );
  expect(container.querySelector('.ui.small.label > i')).toBeInTheDocument();
  expect(screen.getByText('Yes')).toBeInTheDocument();
});

it('test valueToQuery facet function', () => {
  return expect(
    TemporalCoverageFacet.valueToQuery({
      value: [2003, 2005],
      facet: { field: { value: 'temporal' } },
    }),
  ).toEqual({
    i: 'temporal',
    o: 'plone.app.querystring.operation.list.contains',
    v: ['2003', '2004', '2005'],
  });
});

it('renders Slider Ticks', () => {
  const { container } = render(
    <Ticks count={5}>
      {({ ticks }) => (
        <div className="slider-ticks">
          {ticks.map((tick) => (
            <Tick key={tick.id} tick={tick} count={ticks.length} />
          ))}
        </div>
      )}
    </Ticks>,
  );

  expect(container.querySelector('.slider-ticks')).toBeInTheDocument();
  expect(screen.getByText(0)).toHaveStyle('left: 0%');
  expect(screen.getByText(0.2)).toHaveStyle('left: 0.2%');
  expect(screen.getByText(0.4)).toHaveStyle('left: 0.4%');
  expect(screen.getByText(0.6)).toHaveStyle('left: 0.6%');
  expect(screen.getByText(0.8)).toHaveStyle('left: 0.8%');
  expect(screen.getByText(1)).toHaveStyle('left: 1%');
});
