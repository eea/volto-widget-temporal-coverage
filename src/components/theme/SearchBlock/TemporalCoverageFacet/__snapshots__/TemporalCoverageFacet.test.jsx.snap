// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Slider Ticks 1`] = `
<div>
  <div
    class="slider-ticks"
  >
    <div>
      <div
        style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 0%;"
      />
      <div
        style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 0%;"
      >
        0
      </div>
    </div>
    <div>
      <div
        style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 0.2%;"
      />
      <div
        style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 0.2%;"
      >
        0.2
      </div>
    </div>
    <div>
      <div
        style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 0.4%;"
      />
      <div
        style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 0.4%;"
      >
        0.4
      </div>
    </div>
    <div>
      <div
        style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 0.6%;"
      />
      <div
        style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 0.6%;"
      >
        0.6
      </div>
    </div>
    <div>
      <div
        style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 0.8%;"
      />
      <div
        style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 0.8%;"
      >
        0.8
      </div>
    </div>
    <div>
      <div
        style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 1%;"
      />
      <div
        style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 1%;"
      >
        1
      </div>
    </div>
  </div>
</div>
`;

exports[`renders TemporalCoverageFacet 1`] = `
<div>
  <div
    class="temporal-facet"
  >
    <h4
      class="ui header"
    />
    <div
      class="ui form years-input"
    >
      <div
        class="ui input"
      >
        <input
          max="1915"
          min="1890"
          type="number"
          value="1890"
        />
      </div>
      <div
        class="ui input right"
      >
        <input
          max="1915"
          min="1890"
          type="number"
          value="1915"
        />
      </div>
    </div>
    <div
      style="position: relative; width: 100%;"
    >
      <div
        style="position: absolute; transform: translate(0%, -50%); width: 100%; height: 42px; border-radius: 7px; cursor: pointer;"
      />
      <div
        style="position: absolute; width: 100%; height: 14px; transform: translate(0%, -50%); border-radius: 7px; pointer-events: none; background-color: rgb(155, 155, 155);"
      />
      <div
        class="slider-handles"
      >
        <div
          style="left: 0%; position: absolute; transform: translate(-50%, -50%); z-index: 400; width: 26px; height: 42px; cursor: pointer;"
        />
        <div
          aria-valuemax="1915"
          aria-valuemin="1890"
          aria-valuenow="1890"
          role="slider"
          style="left: 0%; position: absolute; transform: translate(-50%, -50%); z-index: 300; width: 24px; height: 24px; border: 0px; border-radius: 50%; box-shadow: 1px 1px 1px 1px rgba(0, 0, 0, 0.2); background-color: rgb(0, 136, 138);"
        />
        <div
          style="left: 100%; position: absolute; transform: translate(-50%, -50%); z-index: 400; width: 26px; height: 42px; cursor: pointer;"
        />
        <div
          aria-valuemax="1915"
          aria-valuemin="1890"
          aria-valuenow="1915"
          role="slider"
          style="left: 100%; position: absolute; transform: translate(-50%, -50%); z-index: 300; width: 24px; height: 24px; border: 0px; border-radius: 50%; box-shadow: 1px 1px 1px 1px rgba(0, 0, 0, 0.2); background-color: rgb(0, 136, 138);"
        />
      </div>
      <div
        class="slider-tracks"
      >
        <div
          style="position: absolute; transform: translate(0%, -50%); height: 14px; z-index: 1; background-color: rgb(0, 84, 138); border-radius: 7px; cursor: pointer; left: 0%; width: 100%;"
        />
      </div>
      <div
        class="slider-ticks"
      >
        <div>
          <div
            style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 0%;"
          />
          <div
            style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 0%;"
          >
            1890
          </div>
        </div>
        <div>
          <div
            style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 20%;"
          />
          <div
            style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 20%;"
          >
            1895
          </div>
        </div>
        <div>
          <div
            style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 40%;"
          />
          <div
            style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 40%;"
          >
            1900
          </div>
        </div>
        <div>
          <div
            style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 60%;"
          />
          <div
            style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 60%;"
          >
            1905
          </div>
        </div>
        <div>
          <div
            style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 80%;"
          />
          <div
            style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 80%;"
          >
            1910
          </div>
        </div>
        <div>
          <div
            style="position: absolute; margin-top: 17px; width: 1px; height: 5px; background-color: rgb(200, 200, 200); left: 100%;"
          />
          <div
            style="position: absolute; margin-top: 25px; font-size: 10px; text-align: center; margin-left: -8.333333333333334%; width: 16.666666666666668%; left: 100%;"
          >
            1915
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders TemporalCoverageFacetFilterListEntry 1`] = `
<div>
  <div
    class="ui small label"
  >
    Yes
    <i
      aria-hidden="true"
      class="delete icon"
    />
  </div>
</div>
`;
